File: python/Bellman_Ford.py
from typing import *

Edges = List[Tuple[int, int, int]]


def Bellman_Ford(es: Edges, V: int, S: int):
    INF = 10**9
    dis = [INF]*V
    dis[S] = 0
    for _ in range(V-1):
        for u, v, w in es:
            if dis[u] == INF:
                continue
            dis[v] = min(dis[v], dis[u]+w)
    for u, v, w in es:
        if dis[u] == INF:
            continue
        if dis[u]+w < dis[v]:
            return []
    return dis

File: cpp/Bellman_Ford.cpp
template <typename T = int>
class Edge
{
public:
    int from, to;
    T cost;
    Edge(int f, int t, T c) : from(f), to(t), cost(c) {}
};

template <typename T>
using Edges = std::vector<Edge<T>>;

template <typename T>
std::vector<T> Bellman_Ford(const Edges<T> &es, int V, int S)
{
    const T INF = std::numeric_limits<T>::max();
    std::vector dis(V, INF);
    dis[S] = 0;
    for (int i = 0; i < V - 1; i++)
    {
        for (auto &e : es)
        {
            if (dis[e.from] == INF)
                continue;
            dis[e.to] = std::min(dis[e.to], dis[e.from] + e.cost);
        }
    }
    for (auto &e : es)
    {
        if (dis[e.from] == INF)
            continue;
        if (dis[e.from] + e.cost < dis[e.to])
        {
            return std::vector<T>();
        }
    }
    return dis;
}